### ISSUE TITLE
Segmentation fault with sub x { shift; @a = @b; eval { use } }

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7082]
Migrated from [rt.perl.org#25824](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=25824) (status was 'resolved')

Searchable as RT25824$

**Comment by p5pRT** [comment]
### From ion@alku.ion.yi.org


This is a bug report for perl from ion\+perlbug@&#8203;ion\.yi\.org\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.




---
$ perl \-e 'sub x \{ shift; @&#8203;a = @&#8203;b; eval \{ use \} \}'
Segmentation fault

To reproduce the segfault the following needs to be done in a sub&#8203;:
\- @&#8203;\_ must be modified\.
\- A list or a hash must be assigned to a list or a hash\.
\- eval \{ use \}; must be done finally\. The parameters for use don't
&nbsp;  seem to matter\.

gdb backtrace&#8203;:
\#0  0x0809e156 in Perl\_pad\_free \(\)
\#1  0x0808e179 in Perl\_op\_clear \(\)
\#2  0x0808e0a8 in Perl\_op\_free \(\)
\#3  0x0808e07b in Perl\_op\_free \(\)
\#4  0x080967b3 in Perl\_newATTRSUB \(\)
\#5  0x0808c297 in Perl\_yyparse \(\)
\#6  0x08061c64 in perl\_parse \(\)
\#7  0x080616cf in perl\_parse \(\)
\#8  0x0805f3da in main \(\)


---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.3&#8203;:

Configured by ion at Mon Jan 19 12&#8203;:00&#8203;:54 EET 2004\.

Summary of my perl5 \(revision 5\.0 version 8 subversion 3\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.0\-gentoo\, archname=i586\-linux
&nbsp;    uname='linux alku 2\.6\.0\-gentoo \#2 smp sun dec 21 10&#8203;:35&#8203;:44 eet 2003 i686 celeron \(mendocino\) genuineintel gnulinux '
&nbsp;    config\_args='\-des \-Darchname=i586\-linux \-Dcccdlflags=\-fPIC \-Dccdlflags=\-rdynamic \-Dcc=gcc \-Dprefix=/usr \-Dvendorprefix=/usr \-Dsiteprefix=/usr \-Dlocincpth=  \-Doptimize=\-O3 \-march=pentium2 \-mcpu=pentium2 \-funroll\-loops \-pipe \-Duselargefiles \-Dd\_dosuid \-Dd\_semctl\_semun \-Dscriptdir=/usr/bin \-Dman3ext=3pm \-Dcf\_by=Gentoo \-Ud\_csh \-Di\_gdbm \-Di\_db \-Di\_ndbm'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-fno\-strict\-aliasing \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O3 \-march=pentium2 \-mcpu=pentium2 \-funroll\-loops \-pipe'\,
&nbsp;    cppflags='\-DPERL5 \-fno\-strict\-aliasing'
&nbsp;    ccversion=''\, gccversion='3\.3\.2 20031218 \(Gentoo Linux 3\.3\.2\-r5\, propolice\-3\.3\-7\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lpthread \-lnsl \-lndbm \-lgdbm \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
&nbsp;    libc=/lib/libc\-2\.3\.3\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.3\.3'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared \-L/usr/local/lib'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.3&#8203;:
&nbsp;    /etc/perl
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.3/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.3
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.3/i586\-linux
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.3
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.2/i586\-linux
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.2
&nbsp;    /usr/lib/perl5/vendor\_perl
&nbsp;    /usr/lib/perl5/5\.8\.3/i586\-linux
&nbsp;    /usr/lib/perl5/5\.8\.3
&nbsp;    /usr/local/lib/site\_perl
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2/i586\-linux
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.2
&nbsp;    \.



---
Environment for perl v5\.8\.3&#8203;:
&nbsp;    HOME=/home/ion
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LC\_CTYPE=fi\_FI\.UTF\-8
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/home/ion/bin&#8203;:/sbin&#8203;:/usr/sbin&#8203;:/usr/local/sbin&#8203;:/opt/sbin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/local/bin&#8203;:/opt/bin&#8203;:/usr/i586\-pc\-linux\-gnu/gcc\-bin/3\.3&#8203;:/usr/X11R6/bin&#8203;:/opt/blackdown\-jdk\-1\.4\.1/bin&#8203;:/opt/blackdown\-jdk\-1\.4\.1/jre/bin&#8203;:/usr/qt/3/bin&#8203;:/usr/kde/3\.1/bin&#8203;:/usr/games/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Feb 02\, 2004 at 03&#8203;:10&#8203;:46AM \-0000\, ion@&#8203;alku\.ion\.yi\.org \(via RT\) wrote&#8203;:
> \# New Ticket Created by  ion@&#8203;alku\.ion\.yi\.org 
> \# Please include the string&#8203;:  \[perl \#25824\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=25824 >
> 
> 
> 
> This is a bug report for perl from ion\+perlbug@&#8203;ion\.yi\.org\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.3\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> $ perl \-e 'sub x \{ shift; @&#8203;a = @&#8203;b; eval \{ use \} \}'
> Segmentation fault

Thanks for the report\.

P5Pers&#8203;: it can be reduced to&#8203;:

&nbsp;    \./perl  \-e 'sub f \{ @&#8203;a=@&#8203;b=@&#8203;c;  \{use\} \}'

What is happening is that when 'use' is seen\, the parser starts a new
sub by calling start\_subparse\(\) :

&nbsp;    use    :    USE startsub
&nbsp;                \{ CvSPECIAL\_on\(PL\_compcv\); /\* It's a BEGIN \{\} \*/ \}
&nbsp;            WORD WORD listexpr ';'
&nbsp;                \{ utilize\($1\, $2\, $4\, $5\, $6\); \}
&nbsp;        ;
&nbsp;    startsub&#8203;:    /\* NULL \*/    /\* start a regular subroutine scope \*/
&nbsp;            \{ $$ = start\_subparse\(FALSE\, 0\); \}

This creates a new pad and updates PL\_comppad\, PL\_curpad \(with the old
ones being put on the save stack\)\.
When the compile blows up due to the syntax error\, the recovery works
it's way down to the newATTRSUB\(\) call associated with f; but because
PL\_error\_count > 0\, newATTRSUB\(\) does an op\_free\(block\)\. This tries
to free all the PADTMPs associated with all the @&#8203;a=@&#8203;b=@&#8203;c ops\, but
still using the new pad allocated for 'use'\. Coredump ensues\.

Clearly the correct fix is for the savestack to be properly popped during
the error recovery\, by I really don't understand how YACC error recovery
works\, so I'm stuck :\-\(

The index returned by start\_subparse\(\) somehow needs to be saved and
used to pop the savestack\.

Dave\.

\-\- 
Red sky at night \- gerroff my land\!
Red sky at morning \- gerroff my land\!
&nbsp;    \-\- old farmers' sayings \#14

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @rgs

Quoting Dave Mitchell \<davem@&#8203;fdisolutions\.com>&#8203;:
> P5Pers&#8203;: it can be reduced to&#8203;:
> 
>     \./perl  \-e 'sub f \{ @&#8203;a=@&#8203;b=@&#8203;c;  \{use\} \}'
> 
> What is happening is that when 'use' is seen\, the parser starts a new
> sub by calling start\_subparse\(\) :
> 
>     use    :    USE startsub
>                 \{ CvSPECIAL\_on\(PL\_compcv\); /\* It's a BEGIN \{\} \*/ \}
>             WORD WORD listexpr ';'
>                 \{ utilize\($1\, $2\, $4\, $5\, $6\); \}
>         ;
>     startsub&#8203;:    /\* NULL \*/    /\* start a regular subroutine scope \*/
>             \{ $$ = start\_subparse\(FALSE\, 0\); \}
> 
> This creates a new pad and updates PL\_comppad\, PL\_curpad \(with the old
> ones being put on the save stack\)\.
> When the compile blows up due to the syntax error\, the recovery works
> it's way down to the newATTRSUB\(\) call associated with f; but because
> PL\_error\_count > 0\, newATTRSUB\(\) does an op\_free\(block\)\. This tries
> to free all the PADTMPs associated with all the @&#8203;a=@&#8203;b=@&#8203;c ops\, but
> still using the new pad allocated for 'use'\. Coredump ensues\.
> 
> Clearly the correct fix is for the savestack to be properly popped during
> the error recovery\, by I really don't understand how YACC error recovery
> works\, so I'm stuck :\-\(

You can't define "properly" properly when there is a syntax error in
the code being parsed\. \(This reminds me about change \#18166 to fix
a similar bug\.\)

> The index returned by start\_subparse\(\) somehow needs to be saved and
> used to pop the savestack\.

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Feb 03\, 2004 at 05&#8203;:08&#8203;:52PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> Quoting Dave Mitchell \<davem@&#8203;fdisolutions\.com>&#8203;:

> > Clearly the correct fix is for the savestack to be properly popped during
> > the error recovery\, by I really don't understand how YACC error recovery
> > works\, so I'm stuck :\-\(
> 
> You can't define "properly" properly when there is a syntax error in
> the code being parsed\. \(This reminds me about change \#18166 to fix
> a similar bug\.\)

Does this mean that the bug is essentially unfixable without completely
replacing the current parser implementation?

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Feb 04\, 2004 at 10&#8203;:34&#8203;:10PM \+0000\, Nicholas Clark wrote&#8203;:
> On Tue\, Feb 03\, 2004 at 05&#8203;:08&#8203;:52PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> > Quoting Dave Mitchell \<davem@&#8203;fdisolutions\.com>&#8203;:
> 
> > > Clearly the correct fix is for the savestack to be properly popped during
> > > the error recovery\, by I really don't understand how YACC error recovery
> > > works\, so I'm stuck :\-\(
> > 
> > You can't define "properly" properly when there is a syntax error in
> > the code being parsed\. \(This reminds me about change \#18166 to fix
> > a similar bug\.\)
> 
> Does this mean that the bug is essentially unfixable without completely
> replacing the current parser implementation?


No\. As RGS reminded me\, the 'sub;' bug \(\#18166\) was fixable\, so I presume
the 'use;' bug will have a similar fix\. Just haven't looked into it yet\.
\(Too busy installing Fedora Core 1 on my Dell D400 laptop\. Sort of like
trying to get bleedperl to run on an X\-Box\.\)

\-\- 
Justice is when you get what you deserve\.
Law is when you get what you pay for\.

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Feb 04\, 2004 at 10&#8203;:54&#8203;:03PM \+0000\, Dave Mitchell wrote&#8203;:
> On Wed\, Feb 04\, 2004 at 10&#8203;:34&#8203;:10PM \+0000\, Nicholas Clark wrote&#8203;:
> > On Tue\, Feb 03\, 2004 at 05&#8203;:08&#8203;:52PM \+0100\, Rafael Garcia\-Suarez wrote&#8203;:
> > > Quoting Dave Mitchell \<davem@&#8203;fdisolutions\.com>&#8203;:
> > 
> > > > Clearly the correct fix is for the savestack to be properly popped during
> > > > the error recovery\, by I really don't understand how YACC error recovery
> > > > works\, so I'm stuck :\-\(
> > > 
> > > You can't define "properly" properly when there is a syntax error in
> > > the code being parsed\. \(This reminds me about change \#18166 to fix
> > > a similar bug\.\)
> > 
> > Does this mean that the bug is essentially unfixable without completely
> > replacing the current parser implementation?
> 
> 
> No\. As RGS reminded me\, the 'sub;' bug \(\#18166\) was fixable\, so I presume
> the 'use;' bug will have a similar fix\. Just haven't looked into it yet\.
> \(Too busy installing Fedora Core 1 on my Dell D400 laptop\. Sort of like
> trying to get bleedperl to run on an X\-Box\.\)


I've now had time to look at this\. \#18166 wasn't the bug I thought it was;
but regardless\, I think the following patch provides a general solution to
this sort of problem\. At the start of each statement\, we remember the
current value of PL\_savestack\_ix\, and at the end of each statement\, we
pop the stack to that index\. Normally this will be a noop\, but when error
recovery kicks in\, yacc pops tokens until it reaches something that looks
like the end of a statement\.

I'll apply this after I've attempted to sort out run\_byacc\.

Dave\.


\-\- 
Technology is dominated by two types of people&#8203;: those who understand what
they do not manage\, and those who manage what they do not understand\. 

<details><summary>Inline Patch</summary>

```diff
--- ../22243/perly.y	Wed Jan 28 22:05:50 2004
+++ perly.y	Sat Feb  7 16:19:54 2004
@@ -83,7 +83,7 @@
 %token COLONATTR
 
 %type <ival> prog decl format startsub startanonsub startformsub
-%type <ival> progstart remember mremember '&'
+%type <ival> progstart remember mremember '&' savescope
 %type <opval> block mblock lineseq line loop cond else
 %type <opval> expr term subscripted scalar ary hsh arylen star amper sideff
 %type <opval> argexpr nexpr texpr iexpr mexpr mnexpr mtexpr miexpr
@@ -162,16 +162,20 @@
 			{ $$ = block_start(FALSE); }
 	;
 
+savescope:	/* NULL */	/* remember stack pos in case of error */
+		{ $$ = PL_savestack_ix; }
+
 /* A collection of "lines" in the program */
 lineseq	:	/* NULL */
 			{ $$ = Nullop; }
 	|	lineseq decl
 			{ $$ = $1; }
-	|	lineseq line
-			{   $$ = append_list(OP_LINESEQ,
-				(LISTOP*)$1, (LISTOP*)$2);
+	|	lineseq savescope line
+			{   LEAVE_SCOPE($2);
+			    $$ = append_list(OP_LINESEQ,
+				(LISTOP*)$1, (LISTOP*)$3);
 			    PL_pad_reset_pending = TRUE;
-			    if ($1 && $2) PL_hints |= HINT_BLOCK_SCOPE; }
+			    if ($1 && $3) PL_hints |= HINT_BLOCK_SCOPE; }
 	;
 
 /* A "line" in the program */
```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Sat\, Feb 07\, 2004 at 05&#8203;:00&#8203;:19PM \+0000\, Dave Mitchell wrote&#8203;:
> On Wed\, Feb 04\, 2004 at 10&#8203;:54&#8203;:03PM \+0000\, Dave Mitchell wrote&#8203;:
> > No\. As RGS reminded me\, the 'sub;' bug \(\#18166\) was fixable\, so I presume
> > the 'use;' bug will have a similar fix\. Just haven't looked into it yet\.
> > \(Too busy installing Fedora Core 1 on my Dell D400 laptop\. Sort of like
> > trying to get bleedperl to run on an X\-Box\.\)
> 
> 
> I've now had time to look at this\. \#18166 wasn't the bug I thought it was;
> but regardless\, I think the following patch provides a general solution to
> this sort of problem\. At the start of each statement\, we remember the
> current value of PL\_savestack\_ix\, and at the end of each statement\, we
> pop the stack to that index\. Normally this will be a noop\, but when error
> recovery kicks in\, yacc pops tokens until it reaches something that looks
> like the end of a statement\.
> 
> I'll apply this after I've attempted to sort out run\_byacc\.

Having now "sorted out" run\_byacc\, I've applied this patch\. I also
backed out the changes to block\_start\(\) and block\_end\(\) applied in change
\#18166\, since
a\) they relied on a private yacc global variable yynerrs which is no
longer global\,
b\) I think my fix should cover both bugs\.

However\, the new bison parser actually seems to be better at error
recovery\, and in fact both the current bug and the bug originally fixed by
\#18166 have fixed themselves without this new patch\! I still think this
patch is a sensible safeguard against possible future bugs\, so I've
commited it anyway\.

Dave\.

> \-\-\- \.\./22243/perly\.y    Wed Jan 28 22&#8203;:05&#8203;:50 2004
> \+\+\+ perly\.y    Sat Feb  7 16&#8203;:19&#8203;:54 2004
> @&#8203;@&#8203; \-83\,7 \+83\,7 @&#8203;@&#8203;
>  %token COLONATTR
>  
>  %type \<ival> prog decl format startsub startanonsub startformsub
> \-%type \<ival> progstart remember mremember '&'
> \+%type \<ival> progstart remember mremember '&' savescope
>  %type \<opval> block mblock lineseq line loop cond else
>  %type \<opval> expr term subscripted scalar ary hsh arylen star amper sideff
>  %type \<opval> argexpr nexpr texpr iexpr mexpr mnexpr mtexpr miexpr
> @&#8203;@&#8203; \-162\,16 \+162\,20 @&#8203;@&#8203;
>              \{ $$ = block\_start\(FALSE\); \}
>      ;
>  
> \+savescope&#8203;:    /\* NULL \*/    /\* remember stack pos in case of error \*/
> \+        \{ $$ = PL\_savestack\_ix; \}
> \+
>  /\* A collection of "lines" in the program \*/
>  lineseq    :    /\* NULL \*/
>              \{ $$ = Nullop; \}
>      |    lineseq decl
>              \{ $$ = $1; \}
> \-    |    lineseq line
> \-            \{   $$ = append\_list\(OP\_LINESEQ\,
> \-                \(LISTOP\*\)$1\, \(LISTOP\*\)$2\);
> \+    |    lineseq savescope line
> \+            \{   LEAVE\_SCOPE\($2\);
> \+                $$ = append\_list\(OP\_LINESEQ\,
> \+                \(LISTOP\*\)$1\, \(LISTOP\*\)$3\);
>                  PL\_pad\_reset\_pending = TRUE;
> \-                if \($1 && $2\) PL\_hints |= HINT\_BLOCK\_SCOPE; \}
> \+                if \($1 && $3\) PL\_hints |= HINT\_BLOCK\_SCOPE; \}
>      ;
>  
>  /\* A "line" in the program \*/

\-\- 
"Emacs isn't a bad OS once you get used to it\.
It just lacks a decent editor\."

**Comment by p5pRT** [comment]
### From @iabyn

fixed in bleed by change \#22306

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'
